server_test.go

```diff
@@ -992,6 +996,8 @@ func TestServerRoundtripTsig(t *testing.T) {
 			t.Error("missing TSIG")
 		}
 		w.WriteMsg(m)
+
+		close(ch)
```

Comment:
`close(ch)` needs to be moved to the start of this function, otherwise scheduling could cause `Exchange` to return before this line is executed and leave the test flaky.

Reply:
Good catch. Will updare this in a moment.

Reply:
I've simplified this to a boolean. A abusing a channel's state to encode a 2-bit state was a bit overkill.

Reply:
I actually think this needs to stay as a channel. Because the reading and writing are going to happen in separate goroutines with no synchronisation, this is liable to trip the race-detector.

Reply:
Ok, fair enough. I've changed it back to a channel.